### ISSUE TITLE
Inconsistent log positions

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#3651]
**Describe the bug**

During investigation of failed benchmarks I investigated the segmented raft journal and wrote an tool to verify the increasing of the positions and find the lowest and highest position etc. Code can be found here https://github.com/Zelldon/playground/tree/master/zb-log-check.

After copy from one broker the log (+ metadata) and running the tool, I got the following output.

```
Scan log...
===============
At idx 10108
Current position 180358561660928 (Segment id 41993 segment offset 0)
Is smaller then this last position 218180045888864 (Segment id 50799 segment offset 2219360)
Current lowest 218162869953992 current highest 218180045888864
===============
Scan finished
LOG IS INCONSISTENT!
Last position: 180358562692808
Lowest position: 180358561660928
Highest position: 218180045888864
```

As we can see from the output that the position somehow drops from `218180045888864` to `180358561660928`. 

This means on `LogStreamImpl#determineInitialPartitionId` finds the wrong position on restart/leader change. The term of the higher position is 11 and the term of the lower position is 12, which indicates/proofs that this happens on leader change.

Interesting is also that `lastReadAddress()` for position `218180045888864` returns `92682469` and for `180358561660928` `92682471`. It seems that one index is missing?

###  This bug causes

Further investigation showed that seeking to this position is not possible, which can also be a problem on deletion.
For example `reader.seek(218180045888864L);` return false. The problem is that binary search only works if the positions are increasing, since they are not it is not possible to find the right position. `AtomixLogStorageReader#llookUpApproximateAddress(final long position)` will not work as expected. It jumps in the middle and sees that the position is still lower. On next jump it goes over the border where the position dropped, which means it is still lower, it will repeat this until it reaches the end of the log.

### Related to 

It might be related to #3573 but this is not 100% clear, since this benchmark had only one maximum snapshot count.

### Configuration

Configuration for that benchmark was:

```
3 Partitions,
3 Nodes,
3 repl. factor
Default benchmark settings
+ logSegmentSize 128 MB 
+ max snapshot count 1
```

**Comment by Zelldon** [comment]
To check the log later I added an zip
[inconsistentLog.zip](https://github.com/zeebe-io/zeebe/files/4058357/inconsistentLog.zip)

**Comment by Zelldon** [comment]
I scanned the log of the other nodes as well.

**Zeebe-0**
```
Scan log...
===============
At idx 10108
Current position 180358561660928 (Segment id 41993 segment offset 0)
Is smaller then this last position 218180045888864 (Segment id 50799 segment offset 2219360)
Current lowest 218162869953992 current highest 218180045888864
===============
Scan finished
LOG IS INCONSISTENT!
Last position: 180358562692808
Lowest position: 180358561660928
Highest position: 218180045888864
Events: 15070
```

**Zeebe-2**
```
Scan log...
===============
At idx 3061
Current position 180358561660928 (Segment id 41993 segment offset 0)
Is smaller then this last position 218180045888864 (Segment id 50799 segment offset 2219360)
Current lowest 218175754077744 current highest 218180045888864
===============
Scan finished
LOG IS INCONSISTENT!
Last position: 180358562946152
Lowest position: 180358561660928
Highest position: 218180045888864
Events: 9241

```

**Zeebe-1**

```
Scan log...
===============
At idx 2
Current position 180358561660928 (Segment id 41993 segment offset 0)
Is smaller then this last position 218180045888864 (Segment id 50799 segment offset 2219360)
Current lowest 218180045888664 current highest 218180045888864
===============
Scan finished
LOG IS INCONSISTENT!
Last position: 180358562958632
Lowest position: 180358561660928
Highest position: 218180045888864
Events: 6242
```

Interesting is that Zeebe-0, was the leader before and has the longer log. Based on the zbctl output I can see that now Zeebe-2 is the new leader. 

```
[zell classes]$ kubectl exec -it zeebe-0 bash
root@zeebe-0:/usr/local/zeebe# zbctl status --insecure
Cluster size: 3
Partitions count: 3
Replication factor: 3
Brokers:
  Broker 0 - zeebe-0.zeebe.zell-log-128-max-snap-1.svc.cluster.local:26501
    Partition 1 : Follower
    Partition 2 : Follower
    Partition 3 : Follower
  Broker 1 - zeebe-1.zeebe.zell-log-128-max-snap-1.svc.cluster.local:26501
    Partition 1 : Follower
    Partition 2 : Leader
    Partition 3 : Leader
  Broker 2 - zeebe-2.zeebe.zell-log-128-max-snap-1.svc.cluster.local:26501
    Partition 1 : Leader
    Partition 2 : Follower
    Partition 3 : Follower
```


Here you can find the related journal files. 
[inconsistentLog.zip](https://github.com/zeebe-io/zeebe/files/4058660/inconsistentLog.zip)
Weird is that Zeebe-1 has an journal which almost start with the gap and has only the file `raft-partition-partition-1-1.log`.

**Comment by Zelldon** [comment]
Had a look at the benchmark log output and found the following:

```
I 2020-01-10T13:16:07.804778441Z 2020-01-10 13:16:07.804 [] [raft-server-0-raft-partition-partition-1] INFO  io.atomix.protocols.raft.roles.LeaderAppender - RaftServer{raft-partition-partition-1} - ConfigureRequest{term=6, leader=0, index=0, timestamp=1578571491868, members=[DefaultRaftMember{id=2, type=ACTIVE, updated=2020-01-09T12:04:51.868796Z}, DefaultRaftMember{id=0, type=BOOTSTRAP, updated=2020-01-09T12:04:51.868796Z}, DefaultRaftMember{id=1, type=ACTIVE, updated=2020-01-09T12:04:51.868796Z}]} to 2 failed: java.util.concurrent.TimeoutException: Request type raft-partition-partition-1-configure timed out in 5000 milliseconds
 
I 2020-01-10T13:16:07.928788139Z 2020-01-10 13:16:07.928 [partition-1-processor] [Broker-0-zb-actors-2] ERROR io.zeebe.util.actor - Actor failed in phase 'STARTED'. Continue with next job.
 
I 2020-01-10T13:16:07.928837010Z java.lang.IllegalStateException: Expected to find event with the snapshot position 252063044217832 in log stream, but nothing was found. Failed to recover 'partition-1-processor'.
	at io.zeebe.engine.processor.StreamProcessor.recoverFromSnapshot(StreamProcessor.java:202) ~[zeebe-workflow-engine-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.engine.processor.StreamProcessor.onActorStarted(StreamProcessor.java:117) ~[zeebe-workflow-engine-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) ~[zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
 
I 2020-01-10T13:16:07.928932307Z 2020-01-10 13:16:07.928 [partition-1-processor] [Broker-0-zb-actors-2] DEBUG io.zeebe.logstreams - Closed stream processor controller partition-1-processor.
 
I 2020-01-10T13:16:07.929845947Z 2020-01-10 13:16:07.928 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-3] ERROR io.zeebe.broker.system - Unexpected error on stream processor installation!
 
I 2020-01-10T13:16:07.929872568Z java.lang.IllegalStateException: Expected to find event with the snapshot position 252063044217832 in log stream, but nothing was found. Failed to recover 'partition-1-processor'.
	at io.zeebe.engine.processor.StreamProcessor.recoverFromSnapshot(StreamProcessor.java:202) ~[zeebe-workflow-engine-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.engine.processor.StreamProcessor.onActorStarted(StreamProcessor.java:117) ~[zeebe-workflow-engine-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
 
I 2020-01-10T13:16:07.930099863Z 2020-01-10 13:16:07.929 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-3] ERROR io.zeebe.broker.system - Failed to install leader partition 1
 
I 2020-01-10T13:16:07.930122190Z java.lang.IllegalStateException: Expected to find event with the snapshot position 252063044217832 in log stream, but nothing was found. Failed to recover 'partition-1-processor'.
	at io.zeebe.engine.processor.StreamProcessor.recoverFromSnapshot(StreamProcessor.java:202) ~[zeebe-workflow-engine-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.engine.processor.StreamProcessor.onActorStarted(StreamProcessor.java:117) ~[zeebe-workflow-engine-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:115) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:195) [zeebe-util-0.22.0-SNAPSHOT.jar:0.22.0-SNAPSHOT]
 
I 2020-01-10T13:16:09.548618800Z 2020-01-10 13:16:09.548 [] [raft-server-0-raft-partition-partition-3] INFO  io.atomix.protocols.raft.roles.LeaderAppender - RaftServer{raft-partition-partition-3} - AppendRequest{term=3, leader=0, prevLogIndex=88059082, prevLogTerm=3, entries=0, commitIndex=88059082} to 2 failed: java.util.concurrent.TimeoutException: Request type raft-partition-partition-3-append timed out in 5000 milliseconds
 
```

This is probably related due to the log inconsistency since the binary search will not work in that case. Furthermore the problem with this is that the Node is still leader **BUT** no processing runs at this partition. We need to implement the step down functionallity to handle this case #3499 .

**Comment by Zelldon** [comment]
On a different benchmark, with smaller snapshot interval (5min) I found some interesting logs, which seem to be related.


```

  I 2020-01-09T19:19:45.633 [Broker-0-ZeebePartition-3] [Broker-0-zb-actors-3] DEBUG io.zeebe.broker.system - Installing leader partition service for partition PartitionId{id=3, group=raft-partition}
                            
  I 2020-01-09T19:19:45.643 [Broker-0-ZeebePartition-3] [Broker-0-zb-actors-3] DEBUG io.zeebe.logstreams.snapshot - Available snapshots: [SnapshotImpl{position=65962109270104, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/28022266-2-1578596393431-65962109270104}, SnapshotImpl{position=65193311550848, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/27696271-2-1578596079712-65193311550848}, SnapshotImpl{position=64454577279144, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/27381666-2-1578595781952-64454577279144}]
                            
  I 2020-01-09T19:19:46.025 [Broker-0-ZeebePartition-3] [Broker-0-zb-actors-3] DEBUG io.zeebe.logstreams.snapshot - Opened database from '/usr/local/zeebe/data/raft-partition/partitions/3/runtime'.
                            
  I 2020-01-09T19:19:46.025 [Broker-0-ZeebePartition-3] [Broker-0-zb-actors-3] DEBUG io.zeebe.logstreams.snapshot - Recovered state from snapshot 'SnapshotImpl{position=65962109270104, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/28022266-2-1578596393431-65962109270104}'
                            
  I 2020-01-09T19:19:46.037 [partition-3-processor] [Broker-0-zb-actors-1] DEBUG io.zeebe.logstreams - Recovering state of partition 3 from snapshot
                            
  I 2020-01-09T19:19:56.185 [partition-3-processor] [Broker-1-zb-actors-3] INFO  io.zeebe.logstreams - Recovered state of partition 3 from snapshot at position 66013654640192
                            
  I 2020-01-09T19:19:56.186 [partition-3-processor-snapshot-director] [Broker-1-zb-actors-0] DEBUG io.zeebe.logstreams.snapshot - The position of the last valid snapshot is '65962109270104'. Taking snapshots beyond this position.
                            
    I 2020-01-09T19:19:56.1 2 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-1] DEBUG io.zeebe.broker.system - Removing leader partition services for partition PartitionId{id=3, group=raft-partition}
                            
  I 2020-01-09T19:19:56.192 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-1] DEBUG io.zeebe.broker.system - Closing exporter-3
                            
  I 2020-01-09T19:19:56.233 [raft-partition-partition-3-appender] [Broker-1-zb-actors-1] ERROR io.zeebe.logstreams - Failed to append block with last event position 4294981600.
                            
  I 2020-01-09T19:19:56.233 [raft-partition-partition-3-appender] [Broker-1-zb-actors-1] ERROR io.zeebe.logstreams - Failed to append block with last event position 4295038616.
                            
  I 2020-01-09T19:19:56.239 [raft-partition-partition-3-appender] [Broker-1-zb-actors-0] ERROR io.zeebe.logstreams - Failed to append block with last event position 4295038816.
                            
  I 2020-01-09T19:19:56.239 [raft-partition-partition-3-appender] [Broker-1-zb-actors-0] ERROR io.zeebe.logstreams - Failed to append block with last event position 4295066832.
                            
  I 2020-01-09T19:19:56.239 [raft-partition-partition-3-appender] [Broker-1-zb-actors-0] ERROR io.zeebe.logstreams - Failed to append block with last event position 4295094848.
                            
  I 2020-01-09T19:19:56.239 [raft-partition-partition-3-appender] [Broker-1-zb-actors-0] ERROR io.zeebe.logstreams - Failed to append block with last event position 4295122864.
                            
  I 2020-01-09T19:19:56.239 [raft-partition-partition-3-appender] [Broker-1-zb-actors-0] ERROR io.zeebe.logstreams - Failed to append block with last event position 4295150880.
                            
  I 2020-01-09T19:19:56.239 [raft-partition-partition-3-appender] [Broker-1-zb-actors-2] ERROR io.zeebe.logstreams - Failed to append block with last event position 4295164888.
                            
  I 2020-01-09T19:19:58.602 [partition-3-processor] [Broker-0-zb-actors-1] INFO  io.zeebe.logstreams - Recovered state of partition 3 from snapshot at position 66013654640192
                            
  I 2020-01-09T19:19:58.604 [partition-3-processor-snapshot-director] [Broker-0-zb-actors-0] DEBUG io.zeebe.logstreams.snapshot - The position of the last valid snapshot is '65962109270104'. Taking snapshots beyond this position.
                            
    I 2020-01-09T19:20:05.9 8 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-2] DEBUG io.zeebe.broker.system - Closed exporter-3 successfully
                            
  I 2020-01-09T19:20:05.958 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-2] DEBUG io.zeebe.broker.system - Closing partition-3-processor-snapshot-director
                            
  I 2020-01-09T19:20:49.504 [partition-3-processor] [Broker-1-zb-actors-3] INFO  io.zeebe.processor - Processor starts reprocessing, until last source event position 66645013890240
                            
  I 2020-01-09T19:20:57.173 [partition-3-processor] [Broker-0-zb-actors-1] INFO  io.zeebe.processor - Processor starts reprocessing, until last source event position 66645013890240
                            
  I 2020-01-09T19:21:00.020 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-1] DEBUG io.zeebe.broker.system - Closed partition-3-processor-snapshot-director successfully
                            
  I 2020-01-09T19:21:00.020 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-1] DEBUG io.zeebe.broker.system - Closing partition-3-processor
                            
  I 2020-01-09T19:21:00.022 [partition-3-processor] [Broker-1-zb-actors-3] DEBUG io.zeebe.logstreams - Closed stream processor controller partition-3-processor.
                            
    I 2020-01-09T19:21:00.0 2 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-3] DEBUG io.zeebe.broker.system - Closed partition-3-processor successfully
                            
  I 2020-01-09T19:21:00.022 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-3] DEBUG io.zeebe.broker.system - Closing io.zeebe.broker.logstreams.LogStreamDeletionService
                            
  I 2020-01-09T19:21:00.022 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-3] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.logstreams.LogStreamDeletionService successfully
                            
  I 2020-01-09T19:21:00.057 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-3] DEBUG io.zeebe.logstreams.snapshot - Closed database from '/usr/local/zeebe/data/raft-partition/partitions/3/runtime'.
                            
  I 2020-01-09T19:21:00.057 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-3] DEBUG io.zeebe.broker.system - Installing follower partition service for partition PartitionId{id=3, group=raft-partition}
                            
  I 2020-01-09T19:21:00.091 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-3] DEBUG io.zeebe.logstreams.snapshot - Available snapshots: [SnapshotImpl{position=65962109270104, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/28022266-2-1578596393431-65962109270104}, SnapshotImpl{position=65193311550848, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/27696271-2-1578596079712-65193311550848}, SnapshotImpl{position=64454577279144, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/27381666-2-1578595781952-64454577279144}]
                            
  I 2020-01-09T19:21:00.538 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-3] DEBUG io.zeebe.logstreams.snapshot - Opened database from '/usr/local/zeebe/data/raft-partition/partitions/3/runtime'.
                            
  I 2020-01-09T19:21:00.538 [Broker-1-ZeebePartition-3] [Broker-1-zb-actors-3] DEBUG io.zeebe.logstreams.snapshot - Recovered state from snapshot 'SnapshotImpl{position=65962109270104, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/28022266-2-1578596393431-65962109270104}'
                            
  I 2020-01-09T19:21:51.026 [Broker-2-ZeebePartition-3] [Broker-2-zb-actors-2] DEBUG io.zeebe.broker.system - Removing leader partition services for partition PartitionId{id=3, group=raft-partition}
                            
  I 2020-01-09T19:21:51.028 [Broker-2-ZeebePartition-3] [Broker-2-zb-actors-2] DEBUG io.zeebe.broker.system - Partition role transitioning from null to FOLLOWER
                            
  I 2020-01-09T19:21:51.028 [Broker-2-ZeebePartition-3] [Broker-2-zb-actors-2] DEBUG io.zeebe.broker.system - Installing follower partition service for partition PartitionId{id=3, group=raft-partition}
                            
  I 2020-01-09T19:21:51.065 [Broker-2-ZeebePartition-3] [Broker-2-zb-actors-2] DEBUG io.zeebe.logstreams.snapshot - Available snapshots: [SnapshotImpl{position=65962109270104, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/28022266-2-1578596393431-65962109270104}, SnapshotImpl{position=65193311550848, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/27696271-2-1578596079712-65193311550848}, SnapshotImpl{position=64454577279144, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/27381666-2-1578595781952-64454577279144}]
                            
  I 2020-01-09T19:21:51.562 [Broker-2-ZeebePartition-3] [Broker-2-zb-actors-2] DEBUG io.zeebe.logstreams.snapshot - Opened database from '/usr/local/zeebe/data/raft-partition/partitions/3/runtime'.
                            
  I 2020-01-09T19:21:51.562 [Broker-2-ZeebePartition-3] [Broker-2-zb-actors-2] DEBUG io.zeebe.logstreams.snapshot - Recovered state from snapshot 'SnapshotImpl{position=65962109270104, path=/usr/local/zeebe/data/raft-partition/partitions/3/snapshots/28022266-2-1578596393431-65962109270104}'
                            
  I 2020-01-09T19:22:29.029 [partition-3-processor] [Broker-0-zb-actors-3] INFO  io.zeebe.processor - Processor finished reprocessing at event position 66645013890240
                            
  I 2020-01-09T19:24:40.219 [] [raft-partition-group-raft-partition-3] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting up to index 31129396
                            
  I 2020-01-09T19:24:45.376 [] [raft-partition-group-raft-partition-3] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Compacting up to index 31129396
                            
  I 2020-01-09T19:24:58.631 [partition-3-processor-snapshot-director] [Broker-0-zb-fs-workers-0] DEBUG io.zeebe.logstreams.snapshot - No changes since last snapshot we will skip snapshot creation. Last valid snapshot position 65962109270104, new lower bound position 717265923144
                            
  I 2020-01-09T19:29:58.649 [partition-3-processor-snapshot-director] [Broker-0-zb-fs-workers-0] DEBUG io.zeebe.logstreams.snapshot - No changes since last snapshot we will skip snapshot creation. Last valid snapshot position 65962109270104, new lower bound position 1803886817880
                            
  I 2020-01-09T19:30:52.246 [partition-3-processor] [Broker-0-zb-actors-2] DEBUG io.zeebe.broker.workflow - Skipping record [key: 6755399456229708, intent: ELEMENT_ACTIVATING, value: {"bpmnProcessId":"benchmark","version":1,"workflowKey":2251799813685249,"workflowInstanceKey":6755399456229708,"elementId":"benchmark","flowScopeKey":-1,"bpmnElementType":"PROCESS","parentWorkflowInstanceKey":-1,"parentElementInstanceKey":-1}] due to step guard; in-memory element is null
                            
  I 2020-01-09T19:30:52.246 [partition-3-processor] [Broker-0-zb-actors-2] DEBUG io.zeebe.broker.workflow - Skipping record [key: 6755399456229714, intent: ELEMENT_ACTIVATING, value: {"bpmnProcessId":"benchmark","version":1,"workflowKey":2251799813685249,"workflowInstanceKey":6755399456229714,"elementId":"benchmark","flowScopeKey":-1,"bpmnElementType":"PROCESS","parentWorkflowInstanceKey":-1,"parentElementInstanceKey":-1}] due to step guard; in-memory element is null
                            
  I 2020-01-09T19:30:52.247 [partition-3-processor] [Broker-0-zb-actors-2] DEBUG io.zeebe.broker.workflow - Skipping record [key: 6755399456229720, intent: ELEMENT_ACTIVATING, value: {"bpmnProcessId":"benchmark","version":1,"workflowKey":2251799813685249,"workflowInstanceKey":6755399456229720,"elementId":"benchmark","flowScopeKey":-1,"bpmnElementType":"PROCESS","parentWorkflowInstanceKey":-1,"parentElementInstanceKey":-1}] due to step guard; in-memory element is null
                            
  I 2020-01-09T19:30:52.247 [partition-3-processor] [Broker-0-zb-actors-2] DEBUG io.zeebe.broker.workflow - Skipping record [key: 6755399456229726, intent: ELEMENT_ACTIVATING, value: {"bpmnProcessId":"benchmark","version":1,"workflowKey":2251799813685249,"workflowInstanceKey":6755399456229726,"elementId":"benchmark","flowScopeKey":-1,"bpmnElementType":"PROCESS","parentWorkflowInstanceKey":-1,"parentElementInstanceKey":-1}] due to step guard; in-memory element is null

```

Really interesting are the last snapshot attempts, where it tries to take the snapshot
`I 2020-01-09T19:24:58.631 [partition-3-processor-snapshot-director] [Broker-0-zb-fs-workers-0] DEBUG io.zeebe.logstreams.snapshot - No changes since last snapshot we will skip snapshot creation. Last valid snapshot position 65962109270104, new lower bound position 717265923144`

` I 2020-01-09T19:29:58.649 [partition-3-processor-snapshot-director] [Broker-0-zb-fs-workers-0] DEBUG io.zeebe.logstreams.snapshot - No changes since last snapshot we will skip snapshot creation. Last valid snapshot position 65962109270104, new lower bound position 1803886817880
`

On first look it is not obvious but both positions are smaller then `65962109270104` and `717265923144` is smaller then `1803886817880`.

Processing has then also problems on finding corresponding events in memory.

**Comment by Zelldon** [comment]
I think I know finally what the problem is.

On partition bootstrap the LogStream is created and on becoming a leader only the related actors, like processor, exporter etc. are created. This makes sense since the log stream has the same life time as the partition.

On LogStream starting the Appender and WriteBuffer are created/opened.
https://github.com/zeebe-io/zeebe/blob/develop/logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamImpl.java#L194

To open the Writebuffer we need to determine the partition ID and scan the log. All nodes will start there write buffer and on leader change the log is not scanned again. This means the next leader will probably start on a lower position. We have currently a test which verifies that our log is consistent between restarts, but this test only checks this for a single node. This issue will only happen in a cluster.
